## How to handling missing values

### Data Set Information:

Several constraints were placed on the selection of these instances from a larger database. In particular, all patients here are females at least 21 years old of Pima Indian heritage. ADAP is an adaptive learning routine that generates and executes digital analogs of perceptron-like devices. It is a unique algorithm; see the paper for details.


### Attribute Information:

1. Number of times pregnant 
2. Plasma glucose concentration a 2 hours in an oral glucose tolerance test 
3. Diastolic blood pressure (mm Hg) 
4. Triceps skin fold thickness (mm) 
5. 2-Hour serum insulin (mu U/ml) 
6. Body mass index (weight in kg/(height in m)^2) 
7. Diabetes pedigree function 
8. Age (years) 
9. Class variable (0 or 1) 

### 1. Checking Missing Values

In [12]:
# Read in the dataset

import pandas as pd
import numpy as np
data = pd.read_csv('pima-indians-diabetes.csv',header=None)
print(data.describe())

          0    1    2    3    4    5      6    7    8
count   769  769  769  769  769  769    769  769  769
unique   18  137   48   52  187  249    518   53    3
top       1   99   70    0    0   32  0.254   22    0
freq    135   17   57  227  374   13      6   72  500


In [13]:
# Check the Null Values

data.isnull().sum()

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
dtype: int64


#### Observation: 
It seems that no Null values in a dataset. Considering sometimes missing data could be labeled as zero, let's check that.


In [14]:
data.head(10)

0                             1  \
0  Number of times pregnant  Plasma glucose concentration   
1                         6                           148   
2                         1                            85   
3                         8                           183   
4                         1                            89   
5                         0                           137   
6                         5                           116   
7                         3                            78   
8                        10                           115   
9                         2                           197   

                          2                            3  \
0  Diastolic blood pressure  Triceps skin fold thickness   
1                        72                           35   
2                        66                           29   
3                        64                            0   
4                        66                           23   
5                        40                           35   
6                        74                            0   
7                        50                           32   
8                         0                            0   
9                        70                           45   

                      4                5                           6     7  \
0  2-Hour serum insulin  Body mass index  Diabetes pedigree function  Ages   
1                     0             33.6                       0.627    50   
2                     0             26.6                       0.351    31   
3                     0             23.3                       0.672    32   
4                    94             28.1                       0.167    21   
5                   168             43.1                       2.288    33   
6                     0             25.6                       0.201    30   
7                    88               31                       0.248    26   
8                     0             35.3                       0.134    29   
9                   543             30.5                       0.158    53   

                8  
0  Class variable  
1               1  
2               0  
3               1  
4               0  
5               1  
6               0  
7               1  
8               0  
9               1

#### Observation: 

- Labeling data as 0 is another method to indicate missing value. For example, ’Diastolic blood pressure ' could not be zero but a missing value. Thus, there are some missing values in this dataset as 0.


### 2. Labeling Missing Values

In [26]:
(data[[4]] == '0').sum()

4    374
dtype: int64

In [29]:
(data[[1,2,3,4,5,6,7]] == '0').sum()

1      5
2     35
3    227
4    374
5     11
6      0
7      0
dtype: int64

#### Observation: 

It can tell that Col 1,2,5 only have little zero value, while Col 3 and 4 almost have half of the zero values. That means we need different strategies and methods to deal with the missing values in various columns. 


### 3. Replacing Missing Values

#### (1) Replace with NaN

In [32]:
# mark zero values as missing or NaN
data[[1,2,3,4,5]] = data[[1,2,3,4,5]].replace('0',np.NaN)
# count the number of NaN values in each column
print(data.isnull().sum())

0      0
1      5
2     35
3    227
4    374
5     11
6      0
7      0
8      0
dtype: int64


In [33]:
data.head(10)

0                             1  \
0  Number of times pregnant  Plasma glucose concentration   
1                         6                           148   
2                         1                            85   
3                         8                           183   
4                         1                            89   
5                         0                           137   
6                         5                           116   
7                         3                            78   
8                        10                           115   
9                         2                           197   

                          2                            3  \
0  Diastolic blood pressure  Triceps skin fold thickness   
1                        72                           35   
2                        66                           29   
3                        64                          NaN   
4                        66                           23   
5                        40                           35   
6                        74                          NaN   
7                        50                           32   
8                       NaN                          NaN   
9                        70                           45   

                      4                5                           6     7  \
0  2-Hour serum insulin  Body mass index  Diabetes pedigree function  Ages   
1                   NaN             33.6                       0.627    50   
2                   NaN             26.6                       0.351    31   
3                   NaN             23.3                       0.672    32   
4                    94             28.1                       0.167    21   
5                   168             43.1                       2.288    33   
6                   NaN             25.6                       0.201    30   
7                    88               31                       0.248    26   
8                   NaN             35.3                       0.134    29   
9                   543             30.5                       0.158    53   

                8  
0  Class variable  
1               1  
2               0  
3               1  
4               0  
5               1  
6               0  
7               1  
8               0  
9               1

#### (2) Delete the missing values

In [34]:
data.dropna(inplace=True)
data.shape

(393, 9)

#### Observation: 

After dropping all the missing value rows, the records of the dataset decrease from 769 to 393.

#### (3) Estimate the missing values

- Imputing: use the estimate model to replace the missing values.

In [49]:
data2= pd.read_csv('pima-indians-diabetes.csv',header=None)
# mark zero values as missing or NaN
data2[[1,2,3,4,5]] = data2[[1,2,3,4,5]].replace('0',np.NaN)

# fill missing values with mean column values
data2.fillna(data2.mean(),inplace=True)
# count the number of Nan values in each column
data2.isnull().sum()

0      0
1      5
2     35
3    227
4    374
5     11
6      0
7      0
8      0
dtype: int64

#### ??? not working? Need to check !

### Reference: 


1. Replace missing values with the mean. For this age data, we assume that missing values are distributed similarly to the values that are present. The formal name for this assumption is Missing Completely at Random (MCAR). In this case, substituting values that represent the existing distribution, such as the mean, is a reasonable approach.

2. Replace missing values with the median. This is another justifiable way to handle missing-at-random data, although note that it gives a different answer. For categorical data, it's also common to use the mode, the most commonly occurring value.

3. Replace missing values with an interpolated estimate. Inspection of the data shows that the values in this column, years seniority, is ordered, greatest to least. This structure can be exploited by interpolating the missing value. This approach is very effective when it is appropriate, usually with time-series data.

4. Replace missing values with a constant. Sometimes missing values are Missing Not at Random (MNAR), that is, that they are missing for some reason that is reflected by what their value should be. For income, we assume that values are missing because those with very high incomes preferred not to state them. In this case, we can make a reasonable guess for what "high" means and fill in the blanks. It will still be inaccurate, but more accurate than it was.

5. Replace missing values using imputation. Imputation is a way of using features to model each other. That way, when one is missing, the others can be used to fill in the blank in a reasonable way. This example shows Multivariate Imputation by Chained Equations (MICE) in action on age, years seniority, and income. This approach is particularly powerful when features are somewhat related. In this example, someone with many years seniority is likely to have a higher age. This means that the two features share information, and that when one is absent the other can be used to estimate it.

6. Replace missing values with a missing rank. Our knowledge of how parking spaces are numbered let us make a guess here. All the space numbers from 1-11 are accounted for, so the missing one might be 12. Reasoning along these lines can be applied when data is given in the form of a rank--each item has a positive integer value, and there are no duplicates and no gaps.

7. Replace missing values with a dummy value and create an indicator variable for "missing." When a missing value really means that the feature is not applicable, then that fact can be highlighted. Filling in a dummy value that is clearly different from actual values, such as a negative rank, is one way to do this. Another is to create a new true/false feature tracking whether the original feature is missing. These two approaches work together very nicely and are both illustrated here.

8. Replace missing values with 0. A missing numerical value can mean zero. In the case of an RSVP, invitees who are not planning to attend sometimes neglect to respond, but guests planning to attend are more likely to. In this case, filling in missing blanks with a zero is reasonable.

9. Replace missing values with 0 and create an indicator variable for "missing." Replacing missing values requires making assumptions. Whenever your confidence in those assumptions is low, it is safer to also create a true/false feature indicating that the value was missing. This allows many algorithms to learn to weight those differently. In the case of RSVPs, creating a missing value column is a way of noting that there is less confidence connected to it than to a zero in which an invitee sent their regrets.

From: https://gallery.cortanaintelligence.com/Experiment/Methods-for-handling-missing-values-1